apps/web/public/sw.js

```diff
@@ -0,0 +1,411 @@
+});
+
+self.addEventListener("fetch", (event) => {
+  //service worker intercepted a fetch call
+  console.log(event.request.url);
+  if (!isAuth(event.request.url))
+    event.respondWith(
+      handleFetch(event.request).then((response) => {
+        // TODO: Remove this
+        if (event.request.url.includes("tinybird")) {
+          console.log(response);
+        }
+        return response;
+      }),
+    );
+});
+
+self.addEventListener("message", (event) => {
+  //message from webpage
+});
+
+const openDB = (callback) => {
+  return new Promise((resolve, reject) => {
+    let req = indexedDB.open(database, dbVersion);
+    req.onerror = (err) => {
+      //could not open db
+      console.warn(err);
+      DB = null;
+      reject(err);
+    };
+    req.onupgradeneeded = (ev) => {
+      let db = ev.target.result;
+      for (const store of ALL_STORES) {
+        const { name, key, indexes } = store;
+        if (!db.objectStoreNames.contains(name)) {
+          const objectStore = db.createObjectStore(name, {
+            keyPath: key,
+          });
+
+          for (const index of indexes) {
+            const { name, property, params } = index;
+            objectStore.createIndex(name, property, params);
+          }
+        }
+      }
+    };
+    req.onsuccess = (ev) => {
+      DB = ev.target.result;
+      console.log("db opened and upgraded as needed");
+      if (callback) {
+        callback();
+      }
+      resolve();
+    };
+  });
+};
+
+async function handleFetch(request) {
+  const { url } = request;
+  if (url.includes("/api/google/labels")) {
+    // this should be called regardless.
+    fetchAndUpdateIDB(request.clone());
+
+    // if the data is within idb
+    if (DB) {
+      const getData = () =>
+        new Promise(async (resolve, reject) => {
+          let tx = DB.transaction(LABEL_STORE.name, "readonly");
+          let store = tx.objectStore(LABEL_STORE.name);
+          let req = store.getAll();
+
+          req.onsuccess = () => resolve(req.result);
+          req.onerror = () => reject(req.error);
+        });
+      try {
+        const result = await getData();
+        if (result.length)
+          return new Response(JSON.stringify({ labels: result }), {
+            status: 200,
+          });
+      } catch (error) {
+        console.log("Error while getting data from indexeddb", error.message);
+        // return new Response(`Error:  + ${error.message}`, {
+        //   status: 500,
+        //   headers: { "Content-Type": "text/plain" },
+        // });
+      }
+    }
+  }
+
+  if (
+    url.includes(".svg") ||
+    url.includes(".jpg") ||
+    url.includes(".png") ||
+    url.includes(".jpeg")
+  ) {
+    const cacheResponse = await caches.match(request);
+    if (cacheResponse) return cacheResponse;
+    const fetchResponse = await fetch(request);
+
+    //save in image cache
+    console.log(`save an IMAGE file ${url}`);
+    const cache = await caches.open(IMAGE_CACHE);
```

Comment:
we save in the image cache, but when do we ever read from it?

Reply:
https://github.com/elie222/inbox-zero/pull/97/files#diff-f50be0345a656c79db8f461fd7398526a2706e339548490178d905c062d3feedR140-R141

Reply:
`const cacheResponse = await caches.match(request);`
   ` if (cacheResponse) return cacheResponse;`